### PR TITLE
Enable Java 8 API desugaring

### LINKED ISSUE #7052: Enable Java 8 library desugaring (Android Studio 4.0)
Unsure if this will be accepted just yet as it requires a major Android Studio bump - probably fine

Desugaring allows for more Java 8 APIs to be usable.

> **Java 8 library desugaring in D8 and R8**
> Android Studio now includes support for using a number of Java 8 language APIs without requiring a minimum API level for your app.
> 
> Through a process called _desugaring_, the DEX compiler, D8, in Android Studio 3.0 and higher already provided substantial support for Java 8 language features (such as lambda expressions, default interface methods, try with resources, and more). In Android Studio 4.0, the desugaring engine has been extended to be able to desugar Java language APIs. This means that you can now include standard language APIs that were available only in recent Android releases (such as `java.util.streams`) in apps that support older versions of Android.
> 
> The following set of APIs is supported in this release:
> 
> * Sequential streams (`java.util.stream`)
> * A subset of `java.time`
> * `java.util.function`
> * Recent additions to `java.util.{Map,Collection,Comparator}`
> * Optionals (`java.util.Optional`, `java.util.OptionalInt` and `java.util.OptionalDouble`) and some other new classes useful with the above APIs
> * Some additions to `java.util.concurrent.atomic` (new methods on `AtomicInteger`, `AtomicLong` and `AtomicReference`)
> * `ConcurrentHashMap` (with bug fixes for Android 5.0)
> 
> To support these language APIs, D8 compiles a separate library DEX file that contains an implementation of the missing APIs and includes it in your app. The desugaring process rewrites your app’s code to instead use this library at runtime.
> 
> To enable support for these language APIs, include the following in your module’s `build.gradle` file:
> 
> ```groovy
> android {
>   defaultConfig {
>     // Required when setting minSdkVersion to 20 or lower
>     multiDexEnabled true
>   }
> 
>   compileOptions {
>     // Flag to enable support for the new language APIs
>     coreLibraryDesugaringEnabled true
>     // Sets Java compatibility to Java 8
>     sourceCompatibility JavaVersion.VERSION_1_8
>     targetCompatibility JavaVersion.VERSION_1_8
>   }
> }
> 
> dependencies {
>   coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.0.4'
> }
> ```

https://developer.android.com/studio/releases

### PR DESCRIPTION
**Pull request by david-allison-1**
## Purpose / Description
Android Studio 4.0 introduced more methods for desugaring: (#7052)

## Fixes
Fixes #7052 

## Approach
* Enable Desugaring
* Remove methods from `FunctionalInterfaces`

## How Has This Been Tested?

Unit Tested

## Learning
* https://developer.android.com/studio/releases#4-0-0-desugar
* https://github.com/google/desugar_jdk_libs/blob/master/CHANGELOG.md

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by mikehardy** [comment]
Nice! Can I ask for one related but unrelated check? Can you try - with this change in - depending on the non-java7 version of google analytics lib?

**Comment by mikehardy** [comment]
Sloppy clicked - did not mean to submit yet:

https://github.com/mikehardy/google-analytics-java#overview

https://github.com/ankidroid/Anki-Android/blob/7523ca86dbce6f36eb8a0cd46cf4a0357f1456ad/AnkiDroid/build.gradle#L247 - 2.0.11 is most recent non-java7

**Review by mikehardy — APPROVED** [review]

**Comment by david-allison-1** [comment]
https://github.com/ankidroid/Anki-Android/pull/7055/commits/212ec8be126048149c73c991bec122d72e38b850 - could you confirm it's correct

**Comment by mikehardy** [comment]
That's perfect! Or at least it is the perfect test. Might still blow up but it's the correct dependency

**Comment by david-allison-1** [comment]
Weird, fails with a compilation error for instrumented tests:

```
> Task :AnkiDroid:compileReleaseJavaWithJavac
/home/travis/build/ankidroid/Anki-Android/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/LocaleSelectionDialog.java:52: error: package java8.util does not exist
import java8.util.Lists;
                 ^
1 error
> Task :AnkiDroid:compileReleaseJavaWithJavac FAILED
```

**Comment by david-allison-1** [comment]
I've fixed one of the errors, but there are others due to transitive dependency updates

```
> Task :AnkiDroid:compileDebugJavaWithJavac
C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\src\main\java\com\ichi2\anki\web\HttpFetcher.java:126: warning: [deprecation] create(MediaType,byte[]) in RequestBody has been deprecated
                requestBuilder.post(RequestBody.create(null, new byte[0]));
                                               ^
error: warnings found and -Werror specified
warnings found and -Werror specified

C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\src\main\java\com\ichi2\libanki\sync\CountingFileRequestBody.java:64: warning: [deprecation] buffer() in BufferedSink has been deprecated
            while ((read = source.read(sink.buffer(), SEGMENT_SIZE)) != -1) {
                                           ^
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 error
2 warnings

> Task :AnkiDroid:compileDebugJavaWithJavac FAILED
```

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/LocaleSelectionDialog.java [hunk 1]
```diff
@@ -198,7 +197,7 @@ public void setLocale(@NonNull Locale locale) {
         }
 
         public LocaleListAdapter(@NonNull Locale[] locales) {
-            mSelectableLocales = Lists.of(locales);
+            mSelectableLocales = new ArrayList<>(Arrays.asList(locales));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/LocaleSelectionDialog.java:200]: Needs reviewer confirmation

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/LocaleSelectionDialog.java:200]: seems fine

**Comment by david-allison-1** [comment]
```

> Task :AnkiDroid:packageDebug
Warning: Type `j$.util.DesugarTimeZone` was not found, it is required for default or static interface methods desugaring of `j$.time.ZoneId j$.$r8$backportedMethods$utility$TimeZone$toZoneId$dispatchHolder.toZoneId(java.util.TimeZone)`
Warning: Type `j$.util.Optional` was not found, it is required for default or static interface methods desugaring of `j$.util.Optional j$.util.OptionalConversions.convert(java.util.Optional)`
Warning: Type `j$.util.OptionalDouble` was not found, it is required for default or static interface methods desugaring of `j$.util.OptionalDouble j$.util.OptionalConversions.convert(java.util.OptionalDouble)`
Warning: Type `j$.util.OptionalLong` was not found, it is required for default or static interface methods desugaring of `j$.util.OptionalLong j$.util.OptionalConversions.convert(java.util.OptionalLong)`
Warning: Type `j$.util.OptionalInt` was not found, it is required for default or static interface methods desugaring of `j$.util.OptionalInt j$.util.OptionalConversions.convert(java.util.OptionalInt)`
Warning: Type `j$.time.ZonedDateTime` was not found, it is required for default or static interface methods desugaring of `j$.time.ZonedDateTime j$.time.TimeConversions.convert(java.time.ZonedDateTime)`
Warning: Type `j$.time.ZoneId` was not found, it is required for default or static interface methods desugaring of `j$.time.ZoneId j$.time.TimeConversions.convert(java.time.ZoneId)`
Warning: Type `j$.time.MonthDay` was not found, it is required for default or static interface methods desugaring of `j$.time.MonthDay j$.time.TimeConversions.convert(java.time.MonthDay)`
Warning: Type `j$.time.Instant` was not found, it is required for default or static interface methods desugaring of `j$.time.Instant j$.time.TimeConversions.convert(java.time.Instant)`
Warning: Type `j$.time.LocalDate` was not found, it is required for default or static interface methods desugaring of `j$.time.LocalDate j$.time.TimeConversions.convert(java.time.LocalDate)`
Warning: Type `j$.time.Duration` was not found, it is required for default or static interface methods desugaring of `j$.time.Duration j$.time.TimeConversions.convert(java.time.Duration)`
Warning: Type `j$.util.DesugarGregorianCalendar` was not found, it is required for default or static interface methods desugaring of `j$.time.ZonedDateTime j$.$r8$backportedMethods$utility$GregorianCalendar$toZonedDateTime$dispatchHolder.toZonedDateTime(java.util.GregorianCalendar)`
Warning: Type `j$.util.DesugarDate` was not found, it is required for default or static interface methods desugaring of `j$.time.Instant j$.$r8$backportedMethods$utility$Date$toInstant$dispatchHolder.toInstant(java.util.Date)`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/concurrent/ConcurrentHashMap.class:
  Type `j$.$r8$wrapper$java$util$function$BiFunction$-V-WRP` was not found, it is required for default or static interface methods desugaring of `java.lang.Object java.util.concurrent.ConcurrentHashMap.compute(java.lang.Object, $-vivified-$.java.util.function.BiFunction)`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/concurrent/ConcurrentHashMap.class:
  Type `j$.$r8$wrapper$java$util$function$BiConsumer$-V-WRP` was not found, it is required for default or static interface methods desugaring of `void java.util.concurrent.ConcurrentHashMap.forEach($-vivified-$.java.util.function.BiConsumer)`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/concurrent/ConcurrentHashMap.class:
  Type `j$.$r8$wrapper$java$util$function$Function$-V-WRP` was not found, it is required for default or static interface methods desugaring of `java.lang.Object java.util.concurrent.ConcurrentHashMap.computeIfAbsent(java.lang.Object, $-vivified-$.java.util.function.Function)`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/concurrent/ConcurrentHashMap$ValuesView.class:
  Type `j$.$r8$wrapper$java$util$Spliterator$-WRP` was not found, it is required for default or static interface methods desugaring of `$-vivified-$.java.util.Spliterator java.util.concurrent.ConcurrentHashMap$ValuesView.spliterator()`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/concurrent/ConcurrentHashMap$ValuesView.class:
  Type `j$.$r8$wrapper$java$util$function$Consumer$-V-WRP` was not found, it is required for default or static interface methods desugaring of `void java.util.concurrent.ConcurrentHashMap$ValuesView.forEach($-vivified-$.java.util.function.Consumer)`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/stream/Collectors$Partition$1.class:
  Type `j$.$r8$wrapper$java$util$stream$Stream$-WRP` was not found, it is required for default or static interface methods desugaring of `$-vivified-$.java.util.stream.Stream java.util.stream.Collectors$Partition$1.parallelStream()`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/stream/Collectors$Partition$1.class:
  Type `j$.$r8$wrapper$java$util$function$Predicate$-V-WRP` was not found, it is required for default or static interface methods desugaring of `boolean java.util.stream.Collectors$Partition$1.removeIf($-vivified-$.java.util.function.Predicate)`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/stream/Collectors$Partition$1.class:
  Type `j$.$r8$wrapper$java$util$function$IntFunction$-V-WRP` was not found, it is required for default or static interface methods desugaring of `java.lang.Object[] java.util.stream.Collectors$Partition$1.toArray($-vivified-$.java.util.function.IntFunction)`
Warning in synthesized for lambda desugaring:
  Type `j$.$r8$wrapper$java$util$function$ToIntFunction$-V-WRP` was not found, it is required for default or static interface methods desugaring of `java.util.Comparator java.util.-$$Lambda$Map$Entry$g8sc1MgjjhwTaK8zHulzMasixMw.thenComparingInt($-vivified-$.java.util.function.ToIntFunction)`
Warning in synthesized for lambda desugaring:
  Type `j$.$r8$wrapper$java$util$function$ToLongFunction$-V-WRP` was not found, it is required for default or static interface methods desugaring of `java.util.Comparator java.util.-$$Lambda$Map$Entry$g8sc1MgjjhwTaK8zHulzMasixMw.thenComparingLong($-vivified-$.java.util.function.ToLongFunction)`
Warning in synthesized for lambda desugaring:
  Type `j$.$r8$wrapper$java$util$function$ToDoubleFunction$-V-WRP` was not found, it is required for default or static interface methods desugaring of `java.util.Comparator java.util.-$$Lambda$Map$Entry$g8sc1MgjjhwTaK8zHulzMasixMw.thenComparingDouble($-vivified-$.java.util.function.ToDoubleFunction)`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/concurrent/ThreadLocalRandom.class:
  Type `j$.$r8$wrapper$java$util$stream$IntStream$-WRP` was not found, it is required for default or static interface methods desugaring of `$-vivified-$.java.util.stream.IntStream java.util.concurrent.ThreadLocalRandom.ints()`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/concurrent/ThreadLocalRandom.class:
  Type `j$.$r8$wrapper$java$util$stream$LongStream$-WRP` was not found, it is required for default or static interface methods desugaring of `$-vivified-$.java.util.stream.LongStream java.util.concurrent.ThreadLocalRandom.longs()`
Warning in /home/travis/.gradle/caches/modules-2/files-2.1/com.android.tools/desugar_jdk_libs/1.0.10/61bd346a9d026c92028eb8c8c29e09368b22633c/desugar_jdk_libs-1.0.10.jar:java/util/concurrent/ThreadLocalRandom.class:
  Type `j$.$r8$wrapper$java$util$stream$DoubleStream$-WRP` was not found, it is required for default or static interface methods desugaring of `$-vivified-$.java.util.stream.DoubleStream java.util.concurrent.ThreadLocalRandom.doubles(long)`
Warning: Type `j$.util.OptionalConversions` was not found, it is required for default or static interface methods desugaring of `java.util.OptionalLong j$.$r8$wrapper$java$util$stream$LongStream$-WRP.reduce(java.util.function.LongBinaryOperator)`
Warning: Type `j$.util.LongSummaryStatisticsConversions` was not found, it is required for default or static interface methods desugaring of `java.util.LongSummaryStatistics j$.$r8$wrapper$java$util$stream$LongStream$-WRP.summaryStatistics()`
Warning: Type `j$.util.DoubleSummaryStatisticsConversions` was not found, it is required for default or static interface methods desugaring of `java.util.DoubleSummaryStatistics j$.$r8$wrapper$java$util$stream$DoubleStream$-WRP.summaryStatistics()`
Warning: Type `j$.util.IntSummaryStatisticsConversions` was not found, it is required for default or static interface methods desugaring of `java.util.IntSummaryStatistics j$.$r8$wrapper$java$util$stream$IntStream$-WRP.summaryStatistics()`
```

**Comment by david-allison-1** [comment]
Let's shelve this one for now - likely an issue with building for lower versions rather than executing the code.

**Comment by mikehardy** [comment]
There are some ecosystem shifts rumbling in other areas - react-native just bumped to minSdkVersion 19 (android 4.4?) as did a related library I work with for a feature it uses, that's the first API the Chromium engine was used in Android IIRC. Nothing concrete planned now but that seems like the next waystation people are using as minSdkVersions ratchet up, in case that has bearing here. For quantitative info:

All Android versions  1,697,801
Android 10  738,844
Android 9    449,777
Android 8.0 140,532
Android 8.1   98,332
Android 7.0   86,328
Android 6.0   65,741
Android 7.1   46,084
Android 5.1   24,103
Android 4.4   17,484
Android 5.0   12,939
Android 11   11,114
Android 4.2   2,778
Android 4.1   2,056
Android 4.3   1,117

We are definitely at the point where dropping API<19 wouldn't materially matter, and release-2.13 wouldn't be a bad release to leave them with, though I'm not itching to do it as it doesn't seem super useful until you get to 21 and can drop multidex.

Android 4.0.3 - 4.0.4 202
Android 1.5                 68
Android 2.3.3 - 2.3.7   40
Android 3.2                  7
Android 2.2                  2
All other Android versions 253

**Pull request closed without merging**